### PR TITLE
fit get status issue

### PR DESCRIPTION
**Pull request by wy65701436**
1, return the right upload offset for client when asks.
2, do not call ResumeBlobUpload on getting status.
3, return 416 rather than 404 on failed to patch chunk blob.
4, add the missing upload close

Signed-off-by: Wang Yan <wangyan@vmware.com>

### CODE REVIEW

**Comment by sudo-bmitch** [comment]
Looking through the logic of what happens if the connection drops and a client retries an upload of the same blob to the same location (including the state), where would that fail? I'm not sure if the state is checked earlier, but if not, it looks like the following would cancel the upload:

https://github.com/distribution/distribution/blob/d3d6014c309e0122d929ca6b7f1441f1dc35f57f/registry/handlers/blobupload.go#L314-L321

Shouldn't that return a 416 with an updated Range and Location header instead, and not cancel the upload?

**Comment by wy65701436** [comment]
> 

In fact, according to this specification, https://github.com/distribution/distribution/blob/main/docs/spec/api.md#get-blob-upload, the client should have a way to get the upload offset through this API.

However, according to the handler dispatcher, any get method with a UUID will be processed first by ResumeBlobUpload, which doesn't make much sense.

1, http servers should not cancel uploads while processing get methods, no matter what the case.
2, think about this situation: the distributed http server restarts during the phase of writing the upload block, but some of it is successfully written to disk. At this point, the client will just retry to upload the same block to the same location. But actually the location has changed and it should be the end + the size of the partial block (successfully written). But the client has no way to know that and has to call the get-blob-upload API to return the new location(offset).

**Comment by sudo-bmitch** [comment]
The path I'm thinking of is:

- PATCH, partial write
- Retry PATCH, 416 response
- Possible fallback to GET/HEAD for state
- PATCH with new offset/state

I think you're implementing:

- PATCH, partial write
- GET/HEAD on any error to get the new offset/state
- PATCH with new offset/state

The risk I'm seeing with the retry is:

- PATCH, partial write
- Retry PATCH, 404 response, upload canceled
- All further requests fail

I can adjust the PATCH logic in my code to avoid any retries, but I think 416 would be a more appropriate response than a 404 to the retry.

**Comment by milosgajdos** [comment]
I think returning `416` on `PATCH` retry seems reasonable 🤔 In fact the spec does [sort of] cover that case

> If the server cannot accept the chunk, a 416 Requested Range Not Satisfiable response will be returned and will include a Range header indicating the current status:

That being said, this PR is headed in the right direction, just feel we should expand it?

**Comment by wy65701436** [comment]
> I think returning `416` on `PATCH` retry seems reasonable 🤔 In fact the spec does [sort of] cover that case
> 
> > If the server cannot accept the chunk, a 416 Requested Range Not Satisfiable response will be returned and will include a Range header indicating the current status:
> 
> That being said, this PR is headed in the right direction, just feel we should expand it?


I agree that we should disable `upload.Cancel(buh)` in the error handling of ResumeBlobUpload. However, I'm not sure if we need to update the response code from 404 to 416, which would be a breaking change for the client?

**Comment by sudo-bmitch** [comment]
> That being said, this PR is headed in the right direction, just feel we should expand it?

Yes, the PR certainly fixes an issue that needs to be fixed. So either expand our separate PR would both be good.

I picked 416 because it's defined in the spec and I think GHCR does this already.

**Comment by milosgajdos** [comment]
> I picked 416 because it's defined in the spec and I think GHCR does this already.

Personally, I'd also go for `416`. Imho, that's the correct behaviour (according to spec, anyways) -- I see nothing wrong with nudging the existing clients towards this path.

**Comment by wy65701436** [comment]
@milosgajdos @sudo-bmitch I've updated the code according to the about conversation, can you help to review it?

**Comment by sudo-bmitch** [comment]
Overall design looks right to me, but I'm not familiar enough with the code base. I'll make an intentionally broken client to do some testing.

**Review by sudo-bmitch — APPROVED** [review]
Gave it a quick test and the changes LGTM.

**Comment by wy65701436** [comment]
ping @milosgajdos @thaJeztah @deleteriousEffect

File: registry/handlers/blobupload.go [hunk 1]
```diff
@@ -104,19 +107,26 @@ func (buh *blobUploadHandler) StartBlobUpload(w http.ResponseWriter, r *http.Req
 // GetUploadStatus returns the status of a given upload, identified by id.
 func (buh *blobUploadHandler) GetUploadStatus(w http.ResponseWriter, r *http.Request) {
 	if buh.Upload == nil {
-		buh.Errors = append(buh.Errors, v2.ErrorCodeBlobUploadUnknown)
-		return
+		blobs := buh.Repository.Blobs(buh)
+		upload, err := blobs.Resume(buh, buh.UUID)
+
```
Comment (milosgajdos) [inline registry/handlers/blobupload.go:112]: can we remove this empty line here? my OCD is hitting me hard with this one

**Review by milosgajdos — APPROVED** [review]
LGTM! @deleteriousEffect PTAL

**Review by heww — APPROVED** [review]

**Pull request merged**